### DISCUSSION TITLE
Override class used in annotations

### DISCUSSION THREAD

**Discussion by EliaOnceAgain** [pydantic/pydantic#8409] [Question]
### Description

How does pydantic locate the class to be used in an annotated field's instantiation, and is there a way to override it at runtime?

I'd like to override one of the imported models that is used as annotation, such that creating an object from any other model that uses it as annotation will use the new model.  

##### models.py
```Python
from pydantic import BaseModel

class A(BaseModel):
    A1: int

class B(BaseModel):
    B1: A
```

##### main.py
```Python
from pydantic import BaseModel
import models

# New class A
class A(BaseModel):
    A1: int
    A2: int
models.A = A

# Create object of type B
models.B.model_validate({"B1": {"A1": 1, "A2": 2}})

# Actual Output:  B1=A(A1=1)
# Desired Output: B1=A(A1=1, A2=2)
```

### Version

```Text
pydantic version: 2.5.2
        pydantic-core version: 2.14.5
          pydantic-core build: profile=release pgo=true
                 install path: /tmp/test/venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
                     platform: Linux-5.15.133.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
             related packages: typing_extensions-4.9.0
```

**Comment by haizaar** [comment]
I think the issue is that by the time you finishg importing `models.py`, `B` has been already parsed and "baked" by pydnatic. Have you tried tyoing with `B.model_fields` push the new `A` in?
